Add ERC20 docs (#85)
OpenZeppelin/cairo-contracts: docs/ERC20.md

# ERC20

> The ERC20 token standard is a specification for [fungible tokens](https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens), a type of token where all the units are exactly equal between them. The `ERC20.cairo` contract implements an approximation of [EIP-20](https://eips.ethereum.org/EIPS/eip-20) in Cairo for StarkNet.

Reviewer: ```suggestion
The ERC20 token standard is a specification for [fungible tokens](https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens), a type of token where all the units are exactly equal to each other. The `ERC20.cairo` contract implements an approximation of [EIP-20](https://eips.ethereum.org/EIPS/eip-20) in Cairo for StarkNet.
```

## Table of Contents

- [Interface](#interface)
  * [ERC20 compatibility](#erc20-compatibility)
- [Usage](#usage)
- [Extensibility](#extensibility)
- [API Specification](#api-specification)
  * [`name`](#-name-)
  * [`symbol`](#-symbol-)
  * [`decimals`](#-decimals-)
  * [`total_supply`](#-total-supply-)
  * [`balance_of`](#-balance-of-)
  * [`allowance`](#-allowance-)
  * [`transfer`](#-transfer-)
  * [`transfer_from`](#-transfer-from-)
  * [`approve`](#-approve-)

## Interface

```jsx
@contract_interface
namespace IERC20:

> 		func name() -> (res: felt):

Reviewer: ```suggestion
    func name() -> (res: felt):
```

    end

    func symbol() -> (res: felt):
    end

    func decimals() -> (res: felt):
    end

    func total_supply() -> (res: Uint256):
    end

    func balance_of(account: felt) -> (res: Uint256):
    end

    func allowance(owner: felt, spender: felt) -> (res: Uint256):
    end

    func transfer(recipient: felt, amount: Uint256) -> (success: felt):
    end

    func transfer_from(
            sender: felt, 
            recipient: felt, 
            amount: Uint256
        ) -> (success: felt):
    end

    func approve(spender: felt, amount: Uint256) -> (success: felt):
    end
end
```

### ERC20 compatibility

> Although StarkNet is not EVM compatible, this implementation aims to be as close as possible to the ERC20 standard. This is why:

Reviewer: ```suggestion
Although StarkNet is not EVM compatible, this implementation aims to be as close as possible to the ERC20 standard, in the following ways:
```

- it uses Cairo's `uint256` instead of `felt`
- it returns `1` as success to imitate a `bool`
- it makes use of Cairo's short strings to simulate `name` and `symbol`
- it will emit events once they're implemented on StarkNet.

> Notice that `get_decimals` returns a 252-bit `felt`, meaning it can be much larger than the standard's 8-bit `uint8` and this could potentially affect bridging tokens.

Reviewer: This is what I think the policy should be:

```suggestion
Notice that `decimals` returns a 252-bit `felt`, meaning it can be much larger than the standard's 8-bit `uint8`. However, compliant implementations should not return a value that is not representable in `uint8`.
```

Are there other differences with ERC20? At least from this document it looks like we're still using snake case? I would consider adding a list of how this implementation _differs_ from ERC20, to complement the above list about how it's similar.

## Usage

Use cases go from medium of exchange currency to voting rights, staking, and more.

Considering that the constructor method looks like this:

```python
func constructor(
    name: felt,     # Token name as Cairo short string
    symbol: felt,   # Token symbol as Cairo short string
    recipient: felt # Address where to send initial supply to
):
```

> To create a token your need to deploy it like this:

Reviewer: ```suggestion
To create a token you need to deploy it like this:
```

```python
erc20 = await starknet.deploy(
    "contracts/token/ERC20.cairo",
    constructor_calldata=[
        str_to_felt("Token"),    # name
        str_to_felt("TKN"),      # symbol
        account.contract_address # initial supply recipient
    ]
)
```

> Note that we haven't specified the initial supply. This is because it's fixed at `1000`, as well as `decimals` which is fixed at `18`. If you wish to change any of these values, you should modify the constructor.
> 

As most StarkNet contracts, it expects to be called by another contract and it identifies it through `get_caller_address` (analogous to Solidity's `msg.this`). This is why we need an Account contract to interact with it. For example:

```python

> signer = Signer(123456789987654321)

Reviewer: I still think this makes for confusing documentation. If it was something like `Signer.fromPrivateKey(1234...)` it would be clearer. Otherwise `Signer(PRIVATE_KEY)` might make it clearer, there is no need to define the variable even IMO.

For example, you could:

- Implement a pausing mechanism
- Add roles such as owner or minter
- Modify the `_transfer` function to perform actions [before](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L229) or after [transfers](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L240)

## API Specification

```jsx
func name() -> (res: felt):
end

func symbol() -> (res: felt):
end

func decimals() -> (res: felt):
end

func total_supply() -> (res: Uint256):
end

func balance_of(account: felt) -> (res: Uint256):
end

func allowance(owner: felt, spender: felt) -> (res: Uint256):
end

func transfer(recipient: felt, amount: Uint256) -> (success: felt):
end

func transfer_from(
        sender: felt, 
        recipient: felt, 
        amount: Uint256
    ) -> (success: felt):
end

func approve(spender: felt, amount: Uint256) -> (success: felt):
end
```

### `name`

Returns the name of the token.

Parameters: None.

Returns:

```jsx
name: felt
```

### `symbol`

> Returns the symbol of the token.

Reviewer: ```suggestion
Returns the ticker symbol of the token.
```

Parameters: None.

Returns:

```jsx
symbol: felt
```

### `decimals`

> Returns the decimals places of the token.

Reviewer: Copied from the ERC20 spec:

```suggestion
Returns the number of decimals the token uses - e.g. 8 means to divide the token amount by 100000000 to get its user representation.
```

Parameters: None.

Returns:

```jsx
decimals: felt
```

### `total_supply`

Returns the amount of tokens in existence.

Parameters: None.

Returns:

```jsx
total_supply: Uint256
```

### `balance_of`

Returns the amount of tokens owned by `account`.

Parameters:

```jsx
account: felt
```

Returns:

```jsx
res: Uint256
```

### `allowance`

Returns the remaining number of tokens that `spender` will be allowed to spend on behalf of `owner` through `transfer_from`. This is zero by default.

This value changes when `approve` or `transfer_from` are called.

Parameters:

```jsx
owner: felt
spender: felt
```

Returns:

```jsx
res: Uint256
```

### `transfer`

> Moves `amount` tokens from the caller’s account to `recipient`. It returns `1` representing a bool if succeeds.

Reviewer: ```suggestion
Moves `amount` tokens from the caller’s account to `recipient`. It returns `1` representing a bool if it succeeds.
```

Parameters:

```jsx
recipient: felt
amount: Uint256
```

Returns:

```jsx
success: felt
```

### `transfer_from`

> Moves `amount` tokens from `sender` to `recipient` using the allowance mechanism. `amount` is then deducted from the caller’s allowance. It returns `1` representing a bool if succeeds.

Reviewer: ```suggestion
Moves `amount` tokens from `sender` to `recipient` using the allowance mechanism. `amount` is then deducted from the caller’s allowance. It returns `1` representing a bool if it succeeds.
```

    func decimals() -> (res: felt):
    end

    func total_supply() -> (res: Uint256):
    end

    func balance_of(account: felt) -> (res: Uint256):
    end

    func allowance(owner: felt, spender: felt) -> (res: Uint256):
    end

    func transfer(recipient: felt, amount: Uint256) -> (success: felt):
    end

    func transfer_from(
            sender: felt, 
            recipient: felt, 
            amount: Uint256
        ) -> (success: felt):
    end

    func approve(spender: felt, amount: Uint256) -> (success: felt):
    end
end
```

### ERC20 compatibility

Although StarkNet is not EVM compatible, this implementation aims to be as close as possible to the ERC20 standard, in the following ways:

- it uses Cairo's `uint256` instead of `felt`
- it returns `1` as success to imitate a `bool`
- it makes use of Cairo's short strings to simulate `name` and `symbol`
- it will emit events once they're implemented on StarkNet.

But some differences can still be found, such as:

- `decimals` returns a 252-bit `felt`, meaning it can be much larger than the standard's 8-bit `uint8`. However, compliant implementations should not return a value that is not representable in `uint8`.

> - `transfer`, `transferFrom` and `approve` will never return anything different from true (`1`) because they will revert on any error

Reviewer: For consistency until we fully switch to camel case:

```suggestion
- `transfer`, `transfer_from` and `approve` will never return anything different from true (`1`) because they will revert on any error
```
